File: src/apple/Packages/MapleUI/Sources/MapleUI/Internal/MuiColorWheelMath.swift
// MuiColorWheelMath.swift — pure hue/saturation↔point math for MuiColorWheel.
// Mirrors the web reference's `mui-color-wheel.component.ts` polar
// convention: hue 0° = right (+x axis), increasing counter-clockwise; puck
// radius = saturation. Kept side-effect-free so it's unit-testable without
// rendering a view or running a gesture.

import Foundation

enum MuiColorWheelMath {
    /// Wraps `deg` into `[0, 360)`. Normalizes `-0` to `0` so a value
    /// comparison never sees a sign-only mismatch (e.g. from `atan2(-0, 1)`
    /// at a pointer exactly on the +x axis).
    static func wrapHue(_ deg: Double) -> Double {
        let wrapped = deg.truncatingRemainder(dividingBy: 360)
        let normalized = wrapped < 0 ? wrapped + 360 : wrapped
        return normalized == 0 ? 0 : normalized
    }

    static func clamp01(_ v: Double) -> Double {
        Swift.min(1, Swift.max(0, v))
    }

    /// Converts a pointer offset from the wheel's center — `dx`/`dy`
    /// normalized to `[-1, 1]` where right/up are positive — into a
    /// hue/saturation pair. A pointer exactly on the center (`dx == dy ==
    /// 0`) keeps `currentHue` rather than producing an undefined angle.
    static func value(dx: Double, dy: Double, currentHue: Double) -> (hue: Double, saturation: Double) {
        let radius = Swift.min(1, (dx * dx + dy * dy).squareRoot())
        let hue = radius == 0 ? currentHue : wrapHue(atan2(dy, dx) * 180 / .pi)
        let roundedHue = hue.rounded().truncatingRemainder(dividingBy: 360)
        return (hue: roundedHue, saturation: (radius * 100).rounded())
    }

    /// Puck position for a given hue/saturation, as box coordinates — `left`
    /// grows right, `top` grows down — the inverse of `value(dx:dy:currentHue:)`.
    static func puckPosition(hue: Double, saturation: Double) -> (leftPct: Double, topPct: Double) {
        let radius = clamp01(saturation / 100)
        let rad = wrapHue(hue) * .pi / 180
        let x = radius * cos(rad)
        let y = radius * sin(rad)
        return (leftPct: (x + 1) * 50, topPct: (1 - y) * 50)
    }
}

File: src/windows/Maple.WinUI/MapleUI/Molecules/MuiColorWheelMath.cs
using System;

namespace Maple.UI
{
    /// <summary>One Color Wheel reading: hue in degrees [0, 360), saturation
    /// in [0, 100].</summary>
    public readonly record struct MuiColorWheelValue(double Hue, double Saturation);

    /// <summary>
    /// Plain, WinUI-free pointer&lt;-&gt;value math behind the Maple.UI
    /// Color Wheel molecule (unified-component-catalog.md §2.1). Same split
    /// as <see cref="MuiSliderMath"/> — linkable into Maple.WinUI.Tests
    /// without a live Window.
    ///
    /// Mirrors the polar convention the Pro Editor's `color-wheel-math.ts`
    /// ships: hue 0° points right and runs counter-clockwise, puck radius
    /// carries saturation, the centre reads neutral (hue 0, saturation 0).
    /// Kept as its own small pure port (matching
    /// `mui-color-wheel.component.ts`'s split from the app component) so the
    /// pointer math is unit-testable without a live Window. This IS the
    /// geometry truth for the wheel — its web sibling had a
    /// puck-50%-too-high bug caught in the parity build.
    /// </summary>
    public static class MuiColorWheelMath
    {
        public static double Clamp01(double v) => Math.Max(0, Math.Min(1, v));

        /// <summary>Wraps any angle into [0, 360).</summary>
        public static double WrapHue(double degrees)
        {
            var wrapped = degrees % 360;
            return wrapped < 0 ? wrapped + 360 : wrapped;
        }

        /// <summary>Converts a pointer position — normalized to the wheel's
        /// bounds, x right and y DOWN in [0, 1] (screen/control coordinates)
        /// — into a (hue, saturation) reading. <paramref name="fallbackHue"/>
        /// is returned for hue when the pointer lands exactly on the
        /// center (radius 0, where angle is undefined) — pass the control's
        /// current hue so a center click doesn't snap hue to 0.</summary>
        public static MuiColorWheelValue PointerToValue(double nx01, double ny01, double fallbackHue)
        {
            var x = Clamp01(nx01) * 2 - 1;
            var y = (1 - Clamp01(ny01)) * 2 - 1; // flip to y-up
            var radius = Math.Min(1, Math.Sqrt(x * x + y * y));
            var hue = radius == 0 ? fallbackHue : WrapHue(Math.Atan2(y, x) * 180 / Math.PI);
            return new MuiColorWheelValue(Math.Round(hue) % 360, Math.Round(radius * 100));
        }

        /// <summary>Converts a (hue, saturation) reading into the puck's
        /// unit position within the wheel — x right, y DOWN in [0, 1], the
        /// same coordinate convention <see cref="PointerToValue"/> takes.</summary>
        public static (double X01, double Y01) ValueToUnitPosition(double hue, double saturation)
        {
            var radius = Clamp01(Math.Clamp(saturation, 0, 100) / 100.0);
            var rad = WrapHue(hue) * Math.PI / 180;
            var px = (radius * Math.Cos(rad) + 1) / 2;
            var py = (radius * Math.Sin(rad) + 1) / 2;
            return (px, 1 - py); // flip back to y-down
        }
    }
}
